tools/dds/dds-sniffer/dds-sniffer.cpp

```diff
@@ -144,159 +159,410 @@ void Sniffer::run( uint32_t seconds )
+    switch( info.status )
     {
-        eprosima::fastdds::dds::TypeSupport m_type( new eprosima::fastrtps::types::DynamicPubSubType( dyn_type ) );
-        m_type.register_type( _participant );
-
-        std::cout << "Discovered type: " << m_type->getName() << " from topic " << topic_name << std::endl;
+    case eprosima::fastrtps::rtps::ReaderDiscoveryInfo::DISCOVERED_READER:
+        if( _print_discoveries )
+        {
+            print_reader_discovered( info, true );
+        }
+        save_topic_reader( info );
+        break;
+    case eprosima::fastrtps::rtps::ReaderDiscoveryInfo::REMOVED_READER:
+        if( _print_discoveries )
+        {
+            print_reader_discovered( info, false );
+        }
+        remove_topic_reader( info );
+        break;
+    }
+}
 
-        // Create a data reader for the topic
-        if( _subscriber == nullptr )
+void dds_sniffer::on_participant_discovery( eprosima::fastdds::dds::DomainParticipant *,
+                                            eprosima::fastrtps::rtps::ParticipantDiscoveryInfo && info )
+{
+    switch( info.status )
+    {
+    case eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::DISCOVERED_PARTICIPANT:
+        if( _print_discoveries )
         {
-            _subscriber = _participant->create_subscriber( eprosima::fastdds::dds::SUBSCRIBER_QOS_DEFAULT, nullptr );
+            print_participant_discovered( info, true );
+        }
+        _discovered_participants[info.info.m_guid] = info.info.m_participantName;
+        break;
+    case eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::REMOVED_PARTICIPANT:
+    case eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::DROPPED_PARTICIPANT:
+        if( _print_discoveries )
+        {
+            print_participant_discovered( info, false );
+        }
+        _discovered_participants.erase( info.info.m_guid );
+        break;
+    }
+}
 
-            if( _subscriber == nullptr )
-            {
-                return;
-            }
+void dds_sniffer::on_type_information_received( eprosima::fastdds::dds::DomainParticipant *,
+                                                const eprosima::fastrtps::string_255 topic_name,
+                                                const eprosima::fastrtps::string_255 type_name,
+                                                const eprosima::fastrtps::types::TypeInformation & type_information )
+{
+    std::cout << "Topic '" << topic_name << "' with type '" << type_name << "' received" << std::endl;
+}
+
+void dds_sniffer::save_topic_writer( const eprosima::fastrtps::rtps::WriterDiscoveryInfo & info )
+{
+    std::string topic_name = info.info.topicName().c_str();
+    _discovered_topics[topic_name].writers.insert( info.info.guid() );
+    save_max_indentation( std::move( topic_name ) );
+}
+
+void dds_sniffer::remove_topic_writer( const eprosima::fastrtps::rtps::WriterDiscoveryInfo & info )
+{
+    auto topic_entry = _discovered_topics[info.info.topicName().c_str()];
```

Comment:
Note that will actually add an entry, as unlikely as it is to happen.
Then you'll have a topic with empty readers & writers -- so maybe take the `if` from line 346 out of the `if` of line 343?

Reply:
I still need to get the topic_info entry to access it's readers and writers. Replaced operator[] access with find